File: cpp/trex_selector_methods/trex_screening/demo_trex_scr_02_mc_sim_screen_trex_mmap/demo_trex_scr_02_mc_sim_screen_trex_mmap.cpp
// ==============================================================================
// demo_trex_scr_02_mc_sim_screen_trex_mmap.cpp
// ==============================================================================
/**
 * @file demo_trex_scr_02_mc_sim_screen_trex_mmap.cpp
 *
 * @brief Screen-TRex Selector — memory-mapped D-file variant.
 *
 * @details
 *  Repeats the Monte Carlo study of demo_trex_scr_01_mc_sim_screen_trex.cpp with
 *  TRexControlParameter::use_memory_mapping = true, so the dummy (D) matrices are
 *  backed by on-disk files instead of RAM.
 *
 *  Monte Carlo comparison: Ordinary vs. Bootstrap-CI Screen-TRex over an SNR
 *  sweep. Simulations are parallelised via OpenMP (run_mc_trials_screen_trex).
 */
// ==============================================================================

// std includes
#include <iostream>
#include <map>
#include <random>
#include <string>
#include <unordered_set>
#include <vector>

// Eigen includes
#include <Eigen/Dense>

// TRex includes
#include <utils/datageneration/utils_datagen.hpp>

// Screen-TRex demo simulation utilities
#include "trex_screening_simulation_utils.hpp"

// ==============================================================================
// Namespace aliases
// ==============================================================================

namespace datagen = trex::utils::datageneration::datagen;
using namespace scr_demo;

// ==============================================================================
// Monte Carlo simulation — memory-mapped
// ==============================================================================

/**
 * @brief SNR-sweep Monte Carlo comparison of Screen-TRex variants (mmap).
 *
 * @details Mirrors demo_trex_scr_01's MC simulation but with
 *          use_memory_mapping = true.  Compares Ordinary Screen-TRex and
 *          Bootstrap-CI Screen-TRex over a range of SNR values.
 *
 * @param num_MC   Number of Monte Carlo runs per (method, SNR) pair.
 * @param high_dim If true, use a high-dimensional setting (n=300, p=1000).
 * @param rnd_coef If true, draw support coefficients from N(0,1) each run.
 */
void demo_ScreenTRex_MMap_MonteCarlo(std::size_t num_MC,
                                      bool        high_dim,
                                      bool        rnd_coef)
{
    std::cout.setf(std::ios::unitbuf);
    cdiag::print_section_header(
        "Demo: Screen-TRex Monte Carlo — Memory-Mapped");

    const std::size_t n = high_dim ? 300  : 1000;
    const std::size_t p = high_dim ? 1000 : 300;
    std::cout << (high_dim ? "High-dimensional (p > n)" : "Low-dimensional (n > p)")
              << "  |  use_memory_mapping = true\n";

    const std::size_t support_size = 10;
    const std::vector<double> snr_values =
        {0.01, 0.1, 0.2, 0.5, 0.6, 1.0, 2.0, 5.0};

    const auto methods = default_methods();

    // Storage: method → SNR-indexed vectors
    std::map<std::string, Eigen::VectorXd> est_fdr_map, fdr_map, tpr_map;
    for (const auto& m : methods) {
        const auto len = static_cast<Eigen::Index>(snr_values.size());
        est_fdr_map[m.name] = Eigen::VectorXd::Zero(len);
        fdr_map[m.name]     = Eigen::VectorXd::Zero(len);
        tpr_map[m.name]     = Eigen::VectorXd::Zero(len);
    }

    auto trex_ctrl = make_trex_control_mmap();

    // -----------------------------------------------------------------------
    // Method loop
    // -----------------------------------------------------------------------
    for (const auto& method : methods) {
        std::cout << std::string(70, '=') << "\n";
        std::cout << "Method: " << method.name << "\n";
        std::cout << std::string(70, '=') << "\n";

        auto screen_ctrl = make_screen_control(method.method, method.bootstrap);

        // SNR loop -----------------------------------------------------------
        for (std::size_t snr_idx = 0; snr_idx < snr_values.size(); ++snr_idx) {
            const double snr = snr_values[snr_idx];

            const auto make_data = [=](unsigned seed) -> ScrDGPData {
                return make_iid_dgp(n, p, support_size, snr, rnd_coef, seed);
            };

            const unsigned base_seed = 24u + static_cast<unsigned>(snr_idx) * 1000u;

            std::ostringstream lbl;
            lbl << method.name << "  SNR=" << std::fixed << std::setprecision(2) << snr;
            const auto result = run_mc_trials_screen_trex(
                num_MC, lbl.str(), make_data, trex_ctrl, screen_ctrl, base_seed);

            const auto idx = static_cast<Eigen::Index>(snr_idx);
            est_fdr_map[method.name](idx) = result.avg_est_fdr;
            fdr_map    [method.name](idx) = result.avg_fdr;
            tpr_map    [method.name](idx) = result.avg_tpr;
        }
    }

    const std::string file_stem =
        "scr_screen_trex_mmap_snr_n" + std::to_string(n)
        + "_p" + std::to_string(p);
    save_and_print_mc_results(num_MC, file_stem, snr_values, methods,
                              fdr_map, tpr_map, est_fdr_map);
}


// ==============================================================================
// Main
// ==============================================================================

int main()
{
    std::cout.setf(std::ios::unitbuf);
    omp_set_num_threads(6);

    // -------------------------------------------------------------------------
    // Monte Carlo SNR sweep (Ordinary vs. Bootstrap-CI, memory-mapped)
    // -------------------------------------------------------------------------
    demo_ScreenTRex_MMap_MonteCarlo(
        200,
        true,
        false
    );

    return 0;
}

File: Python/trex_selector_methods/trex_screening/demo_trex_scr_02_mc_sim_screen_trex_mmap/demo_trex_scr_02_mc_sim_screen_trex_mmap.py
# ==============================================================================
# demo_trex_scr_02_mc_sim_screen_trex_mmap.py
# ==============================================================================
#
# Screen-TRex Selector, demo 02: memory-mapped Monte Carlo variant.
#
# Statistically identical to demo 01, but the dummy (D) matrices are backed by
# on-disk files rather than held in RAM: TRexControlParameter.use_memory_mapping
# = True activates the internal D-mmap + solver-serialization pipeline. This is
# the regime Screen-TRex is built for: p so large that a plain in-core T-Rex run
# is impractical. As in the cpp source, the design matrix X still lives in RAM;
# only the dummy matrices move to disk, and the library manages those temporary
# files in its own scratch directories (nothing to clean up in this script).
#
# Mirrors:
#   cpp/trex_selector_methods/trex_screening/demo_trex_scr_02_mc_sim_screen_trex_mmap/
#       demo_trex_scr_02_mc_sim_screen_trex_mmap.cpp
#
# Demo content (single Monte Carlo SNR sweep, memory-mapped):
#   Ordinary vs. Bootstrap-CI Screen-TRex over an SNR grid with the mmap pipeline
#   enabled - equivalence to demo 01 at a lower memory footprint.
#
# Downscaled from cpp: the cpp demo uses num_MC = 200; committed default is
# smaller here for a practical Python runtime (override with SCR_NUM_MC). Support
# indices are 0-based.
# ==============================================================================

import os
import sys

_THIS_DIR = os.path.dirname(os.path.abspath(__file__))
_PARENT_DIR = os.path.dirname(_THIS_DIR)
for _p in (_THIS_DIR, _PARENT_DIR):
    if _p not in sys.path:
        sys.path.insert(0, _p)

from trex_scr_common import default_scr_methods, run_mc_screen, save_and_print_scr_mc

_OUT_DIR = os.path.join(_THIS_DIR, "simulation_results", "data")

# Committed defaults (cpp uses num_MC = 200; downscaled here for Python runtime).
_NUM_MC      = int(os.environ.get("SCR_NUM_MC", 100))
_NUM_WORKERS = int(os.environ.get("SCR_NUM_WORKERS", 6))

N, P, SUPPORT_SIZE = 300, 1000, 10
SNR_GRID = [0.01, 0.1, 0.2, 0.5, 0.6, 1.0, 2.0, 5.0]


# ==============================================================================
# Monte Carlo SNR sweep - Ordinary vs. Bootstrap-CI (memory-mapped)
# ==============================================================================

def demo_screen_trex_mmap_monte_carlo(num_MC=_NUM_MC, num_workers=_NUM_WORKERS):
    print("\n" + "=" * 70)
    print("Demo: Screen-TRex Monte Carlo - Memory-Mapped")
    print("=" * 70)
    print("High-dimensional (p > n)  |  use_memory_mapping = True")
    print(f"n = {N}, p = {P}, s = {SUPPORT_SIZE}, num_MC = {num_MC}\n")

    methods = default_scr_methods()
    method_names = [m["name"] for m in methods]

    def init():
        return {nm: [0.0] * len(SNR_GRID) for nm in method_names}
    fdr_map, tpr_map, est_fdr_map = init(), init(), init()

    trex_args = {"solver": "TLARS", "use_memory_mapping": True, "K": 20}

    for m in methods:
        print("=" * 70)
        print(f"Method: {m['name']}")
        print("=" * 70)
        screen_args = {"trex_method": m["trex_method"], "bootstrap": m["bootstrap"]}
        for si, snr in enumerate(SNR_GRID):
            dgp_args = {"dgp_kind": "iid", "n": N, "p": P,
                        "support_size": SUPPORT_SIZE, "snr": snr, "rnd_coef": False}
            base_seed = 24 + si * 1000        # mirrors cpp base_seed = 24 + snr_idx * 1000
            res = run_mc_screen(num_MC, f"{m['name']}  SNR={snr:.2f}",
                                dgp_args, trex_args, screen_args,
                                base_seed=base_seed, max_workers=num_workers)
            fdr_map[m["name"]][si] = res["avg_fdr"]
            tpr_map[m["name"]][si] = res["avg_tpr"]
            est_fdr_map[m["name"]][si] = res["avg_est_fdr"]

    file_stem = f"scr_screen_trex_mmap_snr_n{N}_p{P}"
    save_and_print_scr_mc(num_MC, _OUT_DIR, file_stem, SNR_GRID,
                          method_names, fdr_map, tpr_map, est_fdr_map)


# ==============================================================================
# Main
# ==============================================================================

def main():
    demo_screen_trex_mmap_monte_carlo()
    print("\ndemo_trex_scr_02_mc_sim_screen_trex_mmap.py complete.")


if __name__ == "__main__":
    main()
